### PR TITLE
[api-documenter] Read config when generating markdown files

### LINKED ISSUE #4048: [api-documenter] Configure `newLineKind` in api-documentor
## Summary

I have configured using `api-extractor.json` to produce `LF` line endings
But we get an output where the line endings are `CLRF`

> there is a change I am just configuring it wrong so let me know if thatis the case

Configure using `newLineKind` the line endings to be `LF`.
Instead we are genearing `CLRF` line endings

## Repro steps

We have the following config
```
{
	"$schema": "https://developer.microsoft.com/json-schemas/api-extractor/v7/api-extractor.schema.json",
	"mainEntryPointFilePath": "<projectFolder>/dist/dts/index.d.ts",
	"newlineKind": "lf",
	"apiReport": {
		"enabled": true,
		"reportFolder": "<projectFolder>/docs",
		"reportFileName": "api-report.md"
	},
	"docModel": {
		"enabled": true,
		"apiJsonFilePath": "<projectFolder>/dist/<unscopedPackageName>.api.json"
	},
	"dtsRollup": {
		"enabled": true
	},
	"compiler": {
		"tsconfigFilePath": "<projectFolder>/tsconfig.build.json"
	},
	"messages": {
		"extractorMessageReporting": {
			"default": {
				"logLevel": "warning"
			},
			"ae-unresolved-link": {
				"logLevel": "error"
			},
			"ae-different-release-tags": {
				"logLevel": "none"
			}
		}
	}
}
```

We then run these two npm commands
```
    "api:extract": "api-extractor run",
    "api:document": "api-documenter markdown -i dist -o docs/api",
```

  **Expected result:** 

File with `LF` line endings

  **Actual result:** 

`cat -e <filename>`
```
<!-- Do not edit this file. It is automatically generated by API Documenter. -->^M$
^M$
[Home](./index.md)^M$
^M$
## API Reference^M$
^M$
## Packages^M$
^M$
|  Package | Description |^M$
|  --- | --- |^M$
|  [package](./index.md) |  |^M$
^M$
```

`^M$` is cat signifying that the file has `CLRF`.

## Standard questions

Please answer these questions to help us investigate your issue more quickly:

| Question | Answer |
| -------- | -------- |
| `@microsoft/api-documenter` version? | 7.21.6 | 
| Operating system? | Mac |
| Documentation target? | Markdown |
| Would you consider contributing a PR? | yes |
| TypeScript compiler version? | 4.5.5 |
| Node.js version (`node -v`)? | v16.17.0 |

### PR DESCRIPTION
**Pull request by matteematt**
Fixing https://github.com/microsoft/rushstack/issues/4048

## Details

Fixes #4048 

## How it was tested

I tested in my repository by using `newLineKind: 'lf'` as the config. This is the only thing I have tested

1. `npm link` in `app/api-documenter`
2. In my local app run `npm link @microsoft/api-documenter`
3. Made my changes on my fork
4. Created the `api-documenter.json` file with the line endings config in my project
5. `$ cd /path/to/matteematt/rushstack && rush rebuild && cd - && npm run api`
6. Observe the markdown generated uses LF line endings now

## Impacted documentation

Committed

### CODE REVIEW

**Comment by matteematt** [comment]
@microsoft-github-policy-service agree

File: apps/api-documenter/src/cli/MarkdownAction.ts [hunk 1]
```diff
@@ -17,12 +20,25 @@ export class MarkdownAction extends BaseAction {
...
+    let configFilePath: string = path.join(process.cwd(), DocumenterConfig.FILENAME);
+
+    // First try the current folder
+    if (!FileSystem.exists(configFilePath)) {
+      // Otherwise try the standard "config" subfolder
+      configFilePath = path.join(process.cwd(), 'config', DocumenterConfig.FILENAME);
+      if (!FileSystem.exists(configFilePath)) {
+        throw new Error(
+          `Unable to find ${DocumenterConfig.FILENAME} in the current folder or in a "config" subfolder`
+        );
```
Comment (iclanton) [inline apps/api-documenter/src/cli/MarkdownAction.ts:32]:
```suggestion
        console.warn(
          `Unable to find ${DocumenterConfig.FILENAME} in the current folder or in a "config" subfolder`
        );
        configFilePath = undefined;
```

File: apps/api-documenter/src/cli/MarkdownAction.ts [hunk 2]
```diff
@@ -17,12 +20,25 @@ export class MarkdownAction extends BaseAction {
...
+      }
+    }
+
+    const documenterConfig: DocumenterConfig = DocumenterConfig.loadFile(configFilePath);
```
Comment (iclanton) [inline apps/api-documenter/src/cli/MarkdownAction.ts:36]:
```suggestion
    const documenterConfig: DocumenterConfig = configFilePath ? DocumenterConfig.loadFile(configFilePath) : undefined;
```

File: common/changes/@microsoft/api-documenter/main_2023-04-04-16-09.json [hunk 3]
```diff
@@ -0,0 +1,10 @@
+{
+  "changes": [
+    {
+      "packageName": "@microsoft/api-documenter",
+      "comment": "[api-documenter] read config file when generating markdown",
+      "type": "patch"
```
Comment (iclanton) [inline common/changes/@microsoft/api-documenter/main_2023-04-04-16-09.json:6]:
```suggestion
      "comment": "Read the `api-documenter.json` config file when generating markdown",
      "type": "minor"
```

**Review by iclanton — APPROVED** [review]

File: apps/api-documenter/src/cli/MarkdownAction.ts [hunk 4]
```diff
@@ -17,12 +20,26 @@ export class MarkdownAction extends BaseAction {
   }
 
   protected async onExecute(): Promise<void> {
-    // override
+    let configFilePath: string = path.join(process.cwd(), DocumenterConfig.FILENAME);
```
Comment (iclanton) [inline apps/api-documenter/src/cli/MarkdownAction.ts:23]:
```suggestion
    let configFilePath: string | undefined = path.join(process.cwd(), DocumenterConfig.FILENAME);
```

File: apps/api-documenter/src/cli/MarkdownAction.ts [hunk 5]
```diff
@@ -17,12 +20,26 @@ export class MarkdownAction extends BaseAction {
...
+      configFilePath = path.join(process.cwd(), 'config', DocumenterConfig.FILENAME);
+      if (!FileSystem.exists(configFilePath)) {
+        console.warn(
+          `Unable to find ${DocumenterConfig.FILENAME} in the current folder or in a "config" subfolder`
+        );
+        configFilePath = undefined;
+      }
+    }
+
+    const documenterConfig: DocumenterConfig = configFilePath ? DocumenterConfig.loadFile(configFilePath) : undefined;
```
Comment (iclanton) [inline apps/api-documenter/src/cli/MarkdownAction.ts:37]:
```suggestion
    const documenterConfig: DocumenterConfig | undefined = configFilePath ? DocumenterConfig.loadFile(configFilePath) : undefined;
```

**Comment by octogonz** [comment]
@matteematt  Sorry to have overlooked this PR earlier.

How I remember it, the goal of `api-documenter generate` was to replace CLI parameters with config file settings (since API Documenter is primarily used in CI pipelines, and it is easier to add code comments and track Git history of a proper config file).  The plan was to eventually deprecate `api-documenter markdown` and `api-documenter yaml`, since choosing the output format would be one of the config file settings.

Adding config file support to `api-documenter markdown` might be an okay feature, but it might also be unnecessary if we can easily solve your problem using `api-documenter generate`. I can take a closer look at this tomorrow. We're going to temporarily hold off publishing this PR until it's decided. Sorry for the confusion!

**Comment by matteematt** [comment]
@octogonz so do you suggest we hold off setting up api extractor to generate with `lf` line endings using my chnage until you've got an outcome with #4189 ?

**Comment by octogonz** [comment]
@matteematt We reviewed this today.

We've decided to revert PR #4049 (and close PR #4189) because you can already solve your problem by using `api-documenter generate` instead of `api-documenter markdown`.  In fact, your PR changes the `api-documenter markdown` command to be functionally equivalent to `api-documenter generate`, with a minor difference that the [outputTarget](https://github.com/microsoft/rushstack/blob/a93a973a9f6c3b6dfdfb15723e29bd024eca6b5c/apps/api-documenter/src/schemas/api-documenter-template.json#L11) setting would be silently ignored (which might be confusing for users).

I apologize for this confusion. It is because `api-documenter generate` was introduced a long time ago as "experimental" but we never got around to documenting it or deprecating the `markdown` and `yaml` actions (because the engineer who introduced it moved on before completing those steps). In the meantime, some new features such as `newlineKind` got introduced which are only available via `api-documenter generate`.

To address this problem, we will do these things:

1. Revert your PR
2. Review the `api-documenter.json` config file design and finalize its JSON schema
3. Graduate `api-documenter generate` to being not experimental
4. Add deprecation notices to `api-documenter markdown` and `api-documenter yaml`
5. Update the website documentation to now instruct people to use `api-documenter generate`

I'll aim to complete it in the next week or so.

CC @iclanton @AlexJerabek @zelliott

**Comment by matteematt** [comment]
@octogonz is there any documentation of what the `api-documenter.json` should look like when using `api-documenter generate`? I am just trying out what we discussed above but I can't quite get the config in a state where it runs the command through

**Comment by octogonz** [comment]
There are a couple test projects in the Rush Stack monorepo, but I don't think we have official docs since it was originally considered "experimental".

**Comment by octogonz** [comment]
Here's a live example: https://github.com/microsoft/rushstack-websites/blob/main/websites/api.rushstack.io/config/api-documenter.json

**Comment by matteematt** [comment]
@octogonz thanks for the reply. I copied your config but I am getting
```
Error loading plugin api-documenter-docusaurus-plugin: Cannot find module 'api-documenter-docusaurus-plugin' from '/path/to/pkg'
npm ERR! Lifecycle script `build` failed with error:
npm ERR! Error: command failed
npm ERR!   in workspace: @genesislcap/foundation-layout@14.64.1
npm ERR!   at location: /path/to/pkg
```
I searched for [npm](https://www.npmjs.com/search?q=api%2Ddocumenter%2Ddocusaurus%2Dplugin) for the package name but it doesn't seem to come up with anything that looks like the exact right thing

**Comment by matteematt** [comment]
Hi @octogonz do you or any of your team have time to answer the above query? Or any updates about generating this document in general? Thanks.

**Comment by matthias-ccri** [comment]
It looks like `api-documenter-docusaurus-plugin` is an internal package that isn't published to npm, so you'll have to create it in your project. Its source can be found here: https://github.com/microsoft/rushstack-websites/tree/443d55340866bfb1d0f04ade883057bbc1e51a6b/tools/api-documenter-docusaurus-plugin

**Comment by dmurvihill** [comment]
I'd like to reopen #4048 please. This PR was reverted because the team wanted to instead graduate and document `api-documenter generate`. However, the latter steps never occurred.

For me, this gave the same result as `api-documenter markdown`, but with Unix line endings:

api-documenter.json
```
{
  "outputTarget": "markdown",
  "newlineKind": "lf"
}
```

`api-documenter generate`

**Pull request merged**